soc/xtensa/intel_adsp/common/clk.c

```diff
@@ -153,3 +153,57 @@ uint32_t adsp_clock_source_frequency(int source)
 
 	return adsp_clk_src_info[source].frequency;
 }
+
+#ifdef CONFIG_ADSP_DYNAMIC_CLOCK_SWITCHING
+/**
+ * @brief Counter for the number of clock frequency changes.
+ *
+ * This variable is used to keep track of the number of times the clock frequency
+ * has been lowered during idle entry. It is incremented when the clock frequency
+ * is reduced and cleared when the frequency is restored upon idle exit.
+ */
+atomic_t clock_change_count;
+
+/**
+ * @brief Reduces the clock frequency when the DSP enters idle state.
+ *
+ * This function is intended to be called when the DSP core is about to enter
+ * an idle state. It checks if the current clock frequency is not already at its
+ * lowest setting and, if not, reduces it to the lowest frequency to save power.
+ * The clock_change_count is incremented to indicate that a frequency change has occurred.
+ *
+ * @see adsp_clock_idle_exit()
+ */
+void adsp_clock_idle_entry(void)
+{
+	/* we are already at the lowest clock, there is no need to do anything */
+	if (platform_cpu_clocks[0].current_freq != platform_cpu_clocks[0].lowest_freq) {
+		select_cpu_clock_hw(platform_cpu_clocks[0].lowest_freq);
+		(void)atomic_inc(&clock_change_count);
+	}
+}
+
+/**
+ * @brief Restores the clock frequency when the DSP exits idle state.
+ *
+ * This function should be called when the DSP core exits an idle state. It checks
+ * if the clock frequency was previously reduced by adsp_clock_idle_entry. If so,
+ * it restores the clock frequency to its previous setting and clears the
+ * clock_change_count to indicate that the clock is back to normal operation frequency.
+ *
+ * @see adsp_clock_idle_entry()
+ */
+void adsp_clock_idle_exit(void)
+{
+	/* clock has not been changed */
+	if (!atomic_get(&clock_change_count))
+		return;
```

Comment:
This atomic is not really protecting this area, but it does not even seen to be necessary, the calls for these functions are already synced.